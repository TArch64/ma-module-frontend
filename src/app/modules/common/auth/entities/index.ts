export { User, IUserJSON } from './user';
export { InsecureUser, IInsecureUserJSON } from './insecure-user';
