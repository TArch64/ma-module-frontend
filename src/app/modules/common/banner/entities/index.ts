export { IBannerOptions, BannerOptions } from './banner-options';
export { BannerRef } from './banner-ref';
export { BannerEventsProxy } from './banner-events-proxy';
