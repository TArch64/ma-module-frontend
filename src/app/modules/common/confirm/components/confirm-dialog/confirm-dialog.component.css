.confirm-dialog__text {
    line-height: 1.43;
    padding: 0;
}

.confirm-dialog__additional-action {
    color: rgba(0, 0, 0, .6);
}

.confirm-dialog__actions {
    display: flex;
    justify-content: flex-end;
    min-height: 0;
    margin-right: calc(-1 * var(--offset-sm));
}

.confirm-dialog__action {
    min-width: 80px;
}
