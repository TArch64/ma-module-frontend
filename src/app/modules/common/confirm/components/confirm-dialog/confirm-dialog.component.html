<div mat-dialog-content>
    <p class="confirm-dialog__text">
        {{ bodyText }}
    </p>

    <mat-checkbox
        class="confirm-dialog__additional-action margin-bottom--xs"
        [formControl]="additionalActionControl!"
        color="primary"
        *ngIf="hasAdditionalAction"
    >
        {{ additionalActionText }}
    </mat-checkbox>
</div>

<div mat-dialog-actions class="confirm-dialog__actions">
    <button
        mat-button
        class="confirm-dialog__action"
        type="button"
        color="primary"
        data-test-selector="decline"
        (click)="decline()"
    >
        {{ declineActionText }}
    </button>

    <button
        mat-raised-button
        class="confirm-dialog__action"
        type="button"
        color="primary"
        cdkFocusInitial
        data-test-selector="confirm"
        (click)="confirm()"
    >
        {{ confirmActionText }}
    </button>
</div>
