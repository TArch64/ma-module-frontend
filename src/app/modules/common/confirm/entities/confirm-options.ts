interface IActionOptions {
    text: string;
}

export interface IConfirmOptions {
    text: string;
    confirmAction?: IActionOptions;
    declineAction?: IActionOptions;
    ignoreDismissEvent?: boolean;
}

export class ConfirmOptions {
    public static create(options: IConfirmOptions): ConfirmOptions {
        return new ConfirmOptions(
            options.text,
            options.confirmAction ?? { text: 'Confirm' },
            options.declineAction ?? { text: 'Cancel' },
            options.ignoreDismissEvent ?? true
        );
    }

    private constructor(
        public readonly text: string,
        public readonly confirmAction: IActionOptions,
        public readonly declineAction: IActionOptions,
        public readonly ignoreDismissEvent: boolean,
    ) {}
}
