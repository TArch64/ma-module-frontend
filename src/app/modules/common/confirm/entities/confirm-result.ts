export class ConfirmResult {
    public static createConfirmed(): ConfirmResult {
        return new ConfirmResult(true);
    }

    public static createDeclined(): ConfirmResult {
        return new ConfirmResult(false);
    }

    private constructor(
        public readonly isConfirmed: boolean
    ) {}
}
