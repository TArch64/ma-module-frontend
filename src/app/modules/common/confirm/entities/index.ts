export {IConfirmOptions, ConfirmOptions} from './confirm-options';
export {ConfirmResult} from './confirm-result';
