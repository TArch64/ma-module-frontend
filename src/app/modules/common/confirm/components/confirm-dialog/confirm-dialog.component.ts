import {Component, Inject} from '@angular/core';
import {ConfirmOptions, ConfirmResult, IAdditionalAction} from "@common/confirm";
import {MAT_DIALOG_DATA, MatDialogRef} from "@angular/material/dialog";
import {FormControl} from "@angular/forms";

@Component({
  selector: 'app-confirm-dialog',
  templateUrl: './confirm-dialog.component.html',
  styleUrls: ['./confirm-dialog.component.css']
})
export class ConfirmDialogComponent {
    public readonly additionalActionControl?: FormControl;

    constructor(
        private readonly dialogRef: MatDialogRef<ConfirmDialogComponent>,
        @Inject(MAT_DIALOG_DATA) private readonly options: ConfirmOptions,
    ) {
        if (this.hasAdditionalAction) {
            this.additionalActionControl = new FormControl(this.options.additionalAction!.initialValue);
        }
    }

    public get bodyText(): string {
        return this.options.text;
    }

    public get confirmActionText(): string {
        return this.options.confirmAction.text;
    }

    public get declineActionText(): string {
        return this.options.declineAction.text;
    }

    public get hasAdditionalAction(): boolean {
        return !!this.options.additionalAction;
    }

    public get additionalActionText(): string | undefined {
        return this.options.additionalAction?.text;
    }

    public confirm(): void {
        const result = ConfirmResult.createConfirmed(this.captureAdditionalValueResult());
        this.complete(result);
    }

    private captureAdditionalValueResult(): IAdditionalAction | null {
        if (!this.hasAdditionalAction) return null;
        return { value: this.additionalActionControl!.value };
    }

    public decline(): void {
        this.complete(ConfirmResult.createDeclined());
    }

    private complete(result: ConfirmResult): void {
        this.dialogRef.close(result);
    }
}
