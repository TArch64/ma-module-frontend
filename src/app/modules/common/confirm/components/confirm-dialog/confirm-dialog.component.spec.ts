import {By} from "@angular/platform-browser";
import {ReactiveFormsModule} from "@angular/forms";
import {MatButtonModule} from "@angular/material/button";
import {MatCheckboxModule} from "@angular/material/checkbox";
import {ComponentFixture, TestBed} from "@angular/core/testing";
import {MAT_DIALOG_DATA, MatDialogRef} from "@angular/material/dialog";
import {ConfirmDialogComponent} from "./confirm-dialog.component";
import {ConfirmOptions} from "../../entities";

class MockMatDialogRef {
    mockClose: jest.SpyInstance<void, []> = jest.spyOn<MockMatDialogRef, 'close'>(this, 'close');

    close(): void {}
}

function createMatDialogRef(): MockMatDialogRef {
    return new MockMatDialogRef();
}

function createComponent(matDialogRef: MockMatDialogRef, options: ConfirmOptions): ComponentFixture<ConfirmDialogComponent> {
    const module = TestBed.configureTestingModule({
        imports: [
            ReactiveFormsModule,
            MatButtonModule,
            MatCheckboxModule
        ],
        declarations: [ConfirmDialogComponent],
        providers: [
            {provide: MatDialogRef, useValue: matDialogRef},
            {provide: MAT_DIALOG_DATA, useValue: options}
        ]
    });
    return module.createComponent(ConfirmDialogComponent);
}

describe('rendering', () => {
    test('should render additional action', () => {
        const options = ConfirmOptions.create({
            text: 'test',
            additionalAction: { text: 'test' }
        })
        const fixture = createComponent(createMatDialogRef(), options);
        fixture.detectChanges();
        const actionEl = fixture.debugElement.query(By.css('.confirm-dialog__additional-action'));

        expect(actionEl).toBeTruthy();
    });

    test('should render additional action', () => {
        const options = ConfirmOptions.create({text: 'test'})
        const fixture = createComponent(createMatDialogRef(), options);
        fixture.detectChanges();
        const actionEl = fixture.debugElement.query(By.css('.confirm-dialog__additional-action'));

        expect(actionEl).toBeFalsy()
    });
});

describe('complete', () => {
    test('should complete with confirm', () => {
        const matDialogRef = createMatDialogRef();
        const options = ConfirmOptions.create({text: 'test'})
        const fixture = createComponent(matDialogRef, options);
        fixture.detectChanges();

        const buttonEl = fixture.debugElement.query(By.css('[data-test-selector="confirm"]'));
        buttonEl.triggerEventHandler('click', null);

        expect(matDialogRef.mockClose.mock.calls).toMatchSnapshot();
    });

    test('should complete with decline', () => {
        const matDialogRef = createMatDialogRef();
        const options = ConfirmOptions.create({text: 'test'})
        const fixture = createComponent(matDialogRef, options);
        fixture.detectChanges();

        const buttonEl = fixture.debugElement.query(By.css('[data-test-selector="decline"]'));
        buttonEl.triggerEventHandler('click', null);

        expect(matDialogRef.mockClose.mock.calls).toMatchSnapshot();
    });

    test('should confirm with additional action', () => {
        const matDialogRef = createMatDialogRef();
        const options = ConfirmOptions.create({
            text: 'test',
            additionalAction: { text: 'test' }
        })
        const fixture = createComponent(matDialogRef, options);
        fixture.detectChanges();

        const buttonEl = fixture.debugElement.query(By.css('[data-test-selector="confirm"]'));
        buttonEl.triggerEventHandler('click', null);

        expect(matDialogRef.mockClose.mock.calls).toMatchSnapshot();
    });
});
