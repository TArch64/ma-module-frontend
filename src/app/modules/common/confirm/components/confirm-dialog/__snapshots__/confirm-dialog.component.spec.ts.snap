// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complete should complete with confirm 1`] = `
Array [
  Array [
    ConfirmResult {
      "additionalAction": null,
      "isConfirmed": true,
    },
  ],
]
`;

exports[`complete should complete with decline 1`] = `
Array [
  Array [
    ConfirmResult {
      "additionalAction": null,
      "isConfirmed": false,
    },
  ],
]
`;

exports[`complete should confirm with additional action 1`] = `
Array [
  Array [
    ConfirmResult {
      "additionalAction": Object {
        "value": true,
      },
      "isConfirmed": true,
    },
  ],
]
`;
